### PR TITLE
HPCC-23130 Re-enable threadLocal option for embedded Java

### PR DESCRIPTION
**Pull request by richardkchapman**
Embedded Java's threadLocal option was disabled in 7.2.52 and upmerged
versions, because it was not working correctly (causing cores) if a thread in
a threadpool was reused (as is common in Roxie).

Correcting this so that threadLocal can be re-enabled requires a change to the
threadTermHook mechanism so that we can distinguish a threadpool end from a
thread about to be killed, and a codegen change so that the ICodeContext can
be passed per call rather than only at function context creation.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
Ran regression suite locally on all java tests.

### CODE REVIEW

File: ecl/hqlcpp/hqlcpp.cpp [hunk 1]
```diff
@@ -12121,6 +12121,9 @@ void HqlCppTranslator::buildScriptFunctionDefinition(BuildCtx &ctx, IHqlExpressi
         createParam.append(",NULL");
     createParam.append(");");
     funcctx.addQuoted(createParam);
+    if (threadlocal)
+        funcctx.addQuoted("EmbedContextBlock __b(__ctx, ctx);");
+    else
     funcctx.addQuoted("EmbedContextBlock __b(__ctx);");
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlcpp.cpp]: minor: indentation

File: plugins/javaembed/javaembed.cpp [hunk 2]
```diff
@@ -49,7 +49,7 @@ static const char *version = "Java Embed Helper 1.0.0";
 //#define TRACE_GLOBALREF
 //#define TRACE_CLASSFILE
 //#define CHECK_JNI
-//#define FORCE_GC
+#define FORCE_GC
```
Comment (ghalliday) [inline plugins/javaembed/javaembed.cpp]: left on by mistake?

File: plugins/javaembed/javaembed.cpp [hunk 3]
```diff
@@ -3271,6 +3284,21 @@ class JavaEmbedImportContext : public CInterfaceOf<IEmbedFunctionContext>
         if (classLoader)
             JNIenv->DeleteGlobalRef(classLoader);
     }
+    virtual void endThread() override
+    {
+        instance = nullptr;
```
Comment (ghalliday) [inline plugins/javaembed/javaembed.cpp]: does this match the lifetime?

File: plugins/javaembed/javaembed.cpp [hunk 4]
```diff
@@ -4598,17 +4640,20 @@ class JavaEmbedImportContext : public CInterfaceOf<IEmbedFunctionContext>
...
 
-static void releaseContext()
+static void releaseContext(bool isPooled)
 {
     if (threadContext)
     {
+        threadContext->endThread();
+        if (!isPooled)
+        {
         delete threadContext;
```
Comment (ghalliday) [inline plugins/javaembed/javaembed.cpp]: indentation...

File: rtl/eclrtl/eclrtl.cpp [hunk 5]
```diff
@@ -353,7 +353,11 @@ RTLUnicodeConverter * queryRTLUnicodeConverter(char const * codepage)
         unicodeConverterMap = new MapStrToUnicodeConverter;
         // Use thread terminator hook to clear them up on thread exit.
         // NB: May need to revisit if not on a jlib Thread.
+        if (!threadHooked)
+        {
         prevThreadTerminator = addThreadTermFunc(clearUnicodeConverterMap);
```
Comment (ghalliday) [inline rtl/eclrtl/eclrtl.cpp]: indentation

**Review by ghalliday — COMMENTED** [review]
@richardkchapman a couple of comments and some indentation issues

File: rtl/eclrtl/eclrtl.cpp (hunk 5 above)
Comment (richardkchapman) [reply rtl/eclrtl/eclrtl.cpp]: Fixed

File: plugins/javaembed/javaembed.cpp (hunk 4 above)
Comment (richardkchapman) [reply plugins/javaembed/javaembed.cpp]: Fixed

File: plugins/javaembed/javaembed.cpp (hunk 2 above)
Comment (richardkchapman) [reply plugins/javaembed/javaembed.cpp]: Yes, fixed

File: ecl/hqlcpp/hqlcpp.cpp (hunk 1 above)
Comment (richardkchapman) [reply ecl/hqlcpp/hqlcpp.cpp]: Fixed

**Pull request merged**